File: skills/gemini-image/scripts/generate_image.py
#!/usr/bin/env python3
"""Generate images using Gemini/Imagen API.

Usage:
    python generate_image.py "A futuristic city at sunset"
    python generate_image.py "Robot with skateboard" --model imagen-4.0-generate-001
    python generate_image.py "Mountain landscape" --aspect 16:9 --size 2K
    python generate_image.py "Portrait" --num 4 --output images/

Requirements:
    pip install google-genai pillow
"""

import argparse
import os
import sys
from pathlib import Path

# Load .env file if present
try:
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:
    pass


def get_client():
    """Get Gemini API client."""
    try:
        from google import genai
    except ImportError:
        print("Error: google-genai not installed. Run: pip install google-genai")
        sys.exit(1)

    api_key = os.environ.get("GOOGLE_API_KEY") or os.environ.get("GEMINI_API_KEY")
    if not api_key:
        print("Error: Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable")
        sys.exit(1)

    return genai.Client(api_key=api_key)


def generate_image(
    prompt: str,
    model: str = "imagen-4.0-generate-001",
    output_dir: str = ".",
    output_name: str = "generated_image",
    aspect_ratio: str = "1:1",
    image_size: str = "1K",
    num_images: int = 1,
    person_generation: str = "allow_adult",
) -> list[str]:
    """Generate images using Imagen or Gemini models.

    Args:
        prompt: Text description of the image
        model: Model ID (imagen-4.0-generate-001, gemini-3-pro-image-preview)
        output_dir: Directory to save images
        output_name: Base name for output files
        aspect_ratio: Aspect ratio (1:1, 16:9, 9:16, etc.)
        image_size: Resolution (1K, 2K, 4K for gemini-3-pro)
        num_images: Number of images to generate (1-4)
        person_generation: Person policy (dont_allow, allow_adult, allow_all)

    Returns:
        List of saved file paths
    """
    from google.genai import types

    client = get_client()
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    saved_files = []

    if model.startswith("imagen"):
        # Use Imagen API
        response = client.models.generate_images(
            model=model,
            prompt=prompt,
            config=types.GenerateImagesConfig(
                number_of_images=num_images,
                aspect_ratio=aspect_ratio,
                image_size=image_size,
                person_generation=person_generation,
            ),
        )

        for i, generated_image in enumerate(response.generated_images):
            suffix = f"_{i}" if num_images > 1 else ""
            filename = output_path / f"{output_name}{suffix}.png"
            with open(filename, "wb") as f:
                f.write(generated_image.image.image_bytes)
            saved_files.append(str(filename))
            print(f"Saved: {filename}")

    elif model.startswith("gemini"):
        # Use Gemini generateContent API for image generation
        config = types.GenerateContentConfig(
            image_config=types.ImageConfig(
                aspect_ratio=aspect_ratio,
                image_size=image_size,
            )
        )

        response = client.models.generate_content(
            model=model,
            contents=prompt,
            config=config,
        )

        if response.candidates and response.candidates[0].content.parts:
            for i, part in enumerate(response.candidates[0].content.parts):
                if hasattr(part, "inline_data") and part.inline_data:
                    suffix = (
                        f"_{i}" if len(response.candidates[0].content.parts) > 1 else ""
                    )
                    filename = output_path / f"{output_name}{suffix}.png"

                    import base64

                    image_data = part.inline_data.data
                    if isinstance(image_data, str):
                        image_data = base64.b64decode(image_data)

                    with open(filename, "wb") as f:
                        f.write(image_data)
                    saved_files.append(str(filename))
                    print(f"Saved: {filename}")
    else:
        print(f"Error: Unknown model {model}")
        sys.exit(1)

    return saved_files


def main():
    parser = argparse.ArgumentParser(
        description="Generate images using Gemini/Imagen API"
    )
    parser.add_argument("prompt", help="Text description of the image to generate")
    parser.add_argument(
        "--model",
        "-m",
        default="gemini-3-pro-image-preview",
        help="Model ID (default: gemini-3-pro-image-preview). Options: gemini-3-pro-image-preview, gemini-2.5-flash-image, imagen-4.0-generate-001",
    )
    parser.add_argument(
        "--output",
        "-o",
        default=".",
        help="Output directory (default: current directory)",
    )
    parser.add_argument(
        "--name",
        "-n",
        default="generated_image",
        help="Base name for output files (default: generated_image)",
    )
    parser.add_argument(
        "--aspect",
        "-a",
        default="1:1",
        help="Aspect ratio (default: 1:1). Options: 1:1, 3:4, 4:3, 9:16, 16:9, 21:9",
    )
    parser.add_argument(
        "--size",
        "-s",
        default="1K",
        help="Image size (default: 1K). Options: 1K, 2K, 4K (4K only for gemini-3-pro)",
    )
    parser.add_argument(
        "--num",
        type=int,
        default=1,
        help="Number of images to generate (1-4, default: 1)",
    )
    parser.add_argument(
        "--person",
        default="allow_adult",
        choices=["dont_allow", "allow_adult", "allow_all"],
        help="Person generation policy (default: allow_adult)",
    )

    args = parser.parse_args()

    try:
        files = generate_image(
            prompt=args.prompt,
            model=args.model,
            output_dir=args.output,
            output_name=args.name,
            aspect_ratio=args.aspect,
            image_size=args.size,
            num_images=args.num,
            person_generation=args.person,
        )
        print(f"\nGenerated {len(files)} image(s)")
    except Exception as e:
        print(f"Error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: skills/gemini-image/scripts/generate_image.js
#!/usr/bin/env node
"use strict";

import { mkdir, writeFile } from "node:fs/promises";
import path from "node:path";
import process from "node:process";
import { parseArgs } from "node:util";

import dotenv from "dotenv";
import { GoogleGenAI, Modality } from "@google/genai";

dotenv.config();

const DEFAULT_MODEL = "gemini-3.1-flash-image-preview";
const DEFAULT_OUTPUT_DIR = "images/";
const DEFAULT_OUTPUT_NAME = "generated_image";
const DEFAULT_ASPECT = "1:1";
const DEFAULT_SIZE = "1K";
const DEFAULT_NUM_IMAGES = 1;
const DEFAULT_PERSON = "allow_adult";

const SUPPORTED_MODELS = [
  "gemini-3.1-flash-image-preview",
  "gemini-3-pro-image-preview",
  "gemini-2.5-flash-image",
  "imagen-4.0-generate-001",
];

const SUPPORTED_PERSON_POLICIES = ["dont_allow", "allow_adult", "allow_all"];

function printUsage() {
  console.log(`Generate images using Gemini/Imagen API.

Usage:
  node scripts/generate_image.js "A futuristic city at sunset"
  node scripts/generate_image.js "Robot with skateboard" --model imagen-4.0-generate-001
  node scripts/generate_image.js "Mountain landscape" --aspect 16:9 --size 2K
  node scripts/generate_image.js "Portrait" --num 4 --output-dir ./my-images/ --name portrait
  node scripts/generate_image.js "Custom folder" --no-timestamp

Options:
  --model, -m         Model ID (default: ${DEFAULT_MODEL})
  --output-dir, -o    Output directory (default: ${DEFAULT_OUTPUT_DIR})
  --name, -n          Base output filename (default: ${DEFAULT_OUTPUT_NAME})
  --no-timestamp      Disable timestamp suffix in filename
  --aspect, -a        Aspect ratio (default: ${DEFAULT_ASPECT})
  --size, -s          Image size (default: ${DEFAULT_SIZE})
  --num               Number of images to generate (1-4, default: ${DEFAULT_NUM_IMAGES})
  --person            Person policy: dont_allow | allow_adult | allow_all (default: ${DEFAULT_PERSON})
  --help, -h          Show this help message

Requirements:
  npm install @google/genai@latest dotenv@latest
`);
}

function getApiKey() {
  const apiKey = process.env.GOOGLE_API_KEY || process.env.GEMINI_API_KEY;
  if (!apiKey) {
    throw new Error("Set GOOGLE_API_KEY or GEMINI_API_KEY environment variable");
  }
  return apiKey;
}

function getBaseFilename(outputName, useTimestamp) {
  if (!useTimestamp) {
    return outputName;
  }

  const now = new Date();
  const yyyy = String(now.getFullYear());
  const mm = String(now.getMonth() + 1).padStart(2, "0");
  const dd = String(now.getDate()).padStart(2, "0");
  const hh = String(now.getHours()).padStart(2, "0");
  const mi = String(now.getMinutes()).padStart(2, "0");
  const ss = String(now.getSeconds()).padStart(2, "0");
  const timestamp = `${yyyy}${mm}${dd}_${hh}${mi}${ss}`;

  return `${outputName}_${timestamp}`;
}

function decodeImageBytes(imageBytes) {
  if (!imageBytes) {
    return null;
  }

  if (typeof imageBytes === "string") {
    return Buffer.from(imageBytes, "base64");
  }

  if (imageBytes instanceof Uint8Array) {
    return Buffer.from(imageBytes);
  }

  if (Array.isArray(imageBytes)) {
    return Buffer.from(imageBytes);
  }

  return null;
}

async function saveImage(outputDir, baseFilename, imageBytes, suffix = "") {
  const filename = path.join(outputDir, `${baseFilename}${suffix}.png`);
  await writeFile(filename, imageBytes);
  console.log(`Saved: ${filename}`);
  return filename;
}

async function generateImage({
  prompt,
  model = DEFAULT_MODEL,
  outputDir = DEFAULT_OUTPUT_DIR,
  outputName = DEFAULT_OUTPUT_NAME,
  aspectRatio = DEFAULT_ASPECT,
  imageSize = DEFAULT_SIZE,
  numImages = DEFAULT_NUM_IMAGES,
  personGeneration = DEFAULT_PERSON,
  useTimestamp = true,
}) {
  const ai = new GoogleGenAI({ apiKey: getApiKey() });

  await mkdir(outputDir, { recursive: true });
  const savedFiles = [];
  const baseFilename = getBaseFilename(outputName, useTimestamp);

  if (model.startsWith("imagen")) {
    const response = await ai.models.generateImages({
      model,
      prompt,
      config: {
        numberOfImages: numImages,
        aspectRatio,
        imageSize,
        personGeneration,
      },
    });

    for (const [index, generatedImage] of (response.generatedImages || []).entries()) {
      const imageBytes = decodeImageBytes(generatedImage?.image?.imageBytes);
      if (!imageBytes) {
        continue;
      }

      const suffix = numImages > 1 ? `_${index}` : "";
      const filePath = await saveImage(outputDir, baseFilename, imageBytes, suffix);
      savedFiles.push(filePath);
    }

    return savedFiles;
  }

  if (model.startsWith("gemini")) {
    const response = await ai.models.generateContent({
      model,
      contents: prompt,
      config: {
        responseModalities: [Modality.IMAGE],
        imageConfig: {
          aspectRatio,
          imageSize,
        },
      },
    });

    const parts = response.candidates?.[0]?.content?.parts || [];
    for (const [index, part] of parts.entries()) {
      const imageBytes = decodeImageBytes(part?.inlineData?.data);
      if (!imageBytes) {
        continue;
      }

      const suffix = parts.length > 1 ? `_${index}` : "";
      const filePath = await saveImage(outputDir, baseFilename, imageBytes, suffix);
      savedFiles.push(filePath);
    }

    return savedFiles;
  }

  throw new Error(`Unknown model ${model}`);
}

function parseCli() {
  const { values, positionals } = parseArgs({
    options: {
      model: { type: "string", short: "m", default: DEFAULT_MODEL },
      "output-dir": { type: "string", short: "o", default: DEFAULT_OUTPUT_DIR },
      name: { type: "string", short: "n", default: DEFAULT_OUTPUT_NAME },
      "no-timestamp": { type: "boolean", default: false },
      aspect: { type: "string", short: "a", default: DEFAULT_ASPECT },
      size: { type: "string", short: "s", default: DEFAULT_SIZE },
      num: { type: "string", default: String(DEFAULT_NUM_IMAGES) },
      person: { type: "string", default: DEFAULT_PERSON },
      help: { type: "boolean", short: "h", default: false },
    },
    allowPositionals: true,
  });

  if (values.help) {
    printUsage();
    process.exit(0);
  }

  if (positionals.length === 0) {
    printUsage();
    process.exit(1);
  }

  const prompt = positionals.join(" ");
  const numImages = Number(values.num);

  if (Number.isNaN(numImages) || numImages < 1 || numImages > 4) {
    throw new Error("--num must be an integer between 1 and 4");
  }

  if (!SUPPORTED_MODELS.includes(values.model)) {
    throw new Error(`Unsupported model '${values.model}'. Use one of: ${SUPPORTED_MODELS.join(", ")}`);
  }

  if (!SUPPORTED_PERSON_POLICIES.includes(values.person)) {
    throw new Error(
      `Unsupported --person '${values.person}'. Use one of: ${SUPPORTED_PERSON_POLICIES.join(", ")}`,
    );
  }

  return {
    prompt,
    model: values.model,
    outputDir: values["output-dir"],
    outputName: values.name,
    aspectRatio: values.aspect,
    imageSize: values.size,
    numImages,
    personGeneration: values.person,
    useTimestamp: !values["no-timestamp"],
  };
}

async function main() {
  try {
    const options = parseCli();
    const files = await generateImage(options);
    console.log(`\nGenerated ${files.length} image(s)`);
  } catch (error) {
    console.error(`Error: ${error instanceof Error ? error.message : String(error)}`);
    process.exit(1);
  }
}

await main();
